### PR TITLE
fix: fix fetchError persisting across different fetchers

### PR DESCRIPTION
**Pull request by dylanowen**
If you construct a `GraphiQL` instance with a bad fetcher `fetchError` gets persisted in `SchemaContextType` forever. You can still fetch a new schema but the `DocExplorer` will always show "Error fetching schema"

### CODE REVIEW

File: packages/graphiql-react/src/schema.tsx
Comment (thomasheyenbrock) [inline packages/graphiql-react/src/schema.tsx:171]: I don't think this comment is necessary, it's very clear what the following line does.
```suggestion
```

File: packages/graphiql/src/components/__tests__/GraphiQL.spec.tsx
Comment (thomasheyenbrock) [inline packages/graphiql/src/components/__tests__/GraphiQL.spec.tsx:101]: The test description sounds like the refetch would be automatically triggered after a fetch error. Can we make it more clear that the refetch has to be triggered "manually" (i.e. providing a different fetcher)

**Review by thomasheyenbrock — CHANGES_REQUESTED** [review]
Thanks so much @dylanowen for catching this! That's indeed a bug that got in with one of the refactorings we did in the past months (most likely my bad 😓 ). The fix itself is perfect, just have some tiny nits!

Can you also add a changeset here that describes the fix? (I just noticed that this is still missing in our `CONTRIBUTING.md` 🙈 ). Just run `yarn changeset` in the repository root and the CLI will walk you through how to create a changeset which is just a markdown file that you need to commit and push to the branch.

**Comment by thomasheyenbrock** [comment]
@dylanowen and you need to sign the CLA in order to become a contributor. Just click on the details link next to the failing GitHub check, that should walk you through everything.

**Review by lwilli — APPROVED** [review]

File: packages/graphiql/src/components/__tests__/GraphiQL.spec.tsx
Comment (dylanowen) [reply packages/graphiql/src/components/__tests__/GraphiQL.spec.tsx:101]: 👍  How does

> should refresh schema with new fetcher after a fetchError

 sound?

**Comment by dylanowen** [comment]
@thomasheyenbrock thanks for the quick review! I'm still working on the CLA but should have that ready soon.

File: packages/graphiql/src/components/__tests__/GraphiQL.spec.tsx
Comment (thomasheyenbrock) [reply packages/graphiql/src/components/__tests__/GraphiQL.spec.tsx]: Sounds good!

**Comment by acao** [comment]
thanks for this! @dylanowen make sure to sign the CLA bot with the email you used to sign the commits, it's very specific about this but won't tell you this is why 😆

**Comment by dylanowen** [comment]
@acao Sorry for the CLA delay but it should be all set now

**Review by acao — APPROVED** [review]
Looks good to merge!

**Comment by pitgrap** [comment]
looks related to https://github.com/graphql/graphiql/pull/2653

**Comment by jonathanawesome** [comment]
@dylanowen I'm working on a small PR (#2778 ) to update some styling that's unrelated to this PR, but getting a failing test that _is_ part of this PR.

```
```

**Comment by dylanowen** [comment]
@jonathanawesome I'm pretty sure this is caused by the long lag time between my PR being tested and me getting my CLA signed. When I opened this PR those selectors existed: https://github.com/graphql/graphiql/blob/48e7710a80739931565e1d472ed38136cc73865f/packages/graphiql/src/components/DocExplorer.tsx#L123 but I didn't realize they were changed before my merge.

**Comment by jonathanawesome** [comment]
ahh...there it is. Thanks for the explanation, makes sense now.

I've got the fix in with #2778.

**Comment by dylanowen** [comment]
Thanks for the quick fix! I was wondering if it'd be better to not depend on the `DocExplorer` for the test so I was working on 
```javascript
    let foundError = null;
    function ErrorIcon() {
      foundError = useSchemaContext({nonNull: true}).fetchError;
      return <div/>;
    }
    const ERROR_PLUGIN = {
      title: 'test',
      icon: ErrorIcon,
      content: () => <div />,
    };
    function firstFetcher() {
      return Promise.reject('Schema Error');
    }
    function secondFetcher() {
      return Promise.resolve(simpleIntrospection);
    }

    // Use a bad fetcher for our initial render
    const { rerender } = render(<GraphiQL fetcher={firstFetcher} plugins={[ERROR_PLUGIN]} />);
    await wait();

    expect(foundError).toBeTruthy()

    // Re-render with valid fetcher
    rerender(<GraphiQL fetcher={secondFetcher} plugins={[ERROR_PLUGIN]} />);
    await wait();

    expect(foundError).not.toBeTruthy();
```
But I'll defer to whatever solution you think is best.

**Comment by jonathanawesome** [comment]
I think you've a good point...the Doc Explorer isn't in the DOM by default anymore (it must have been when you added your test), so I added a fireEvent to the test to ensure that it's there. 

I also added an additional test [closer to the component](https://github.com/graphql/graphiql/pull/2778/commits/ec90db510209d743301cd9daae4e3ad2a3f606ee). 

These are good, the-way-the-user-sees-it tests, so I think they're sufficient.

**Comment by dylanowen** [comment]
Ahhh nice, even better 👍

**Pull request merged**
